flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/LegacyScheduler.java

```diff
@@ -0,0 +1,639 @@
+	 * @param savepointRestoreSettings {@link SavepointRestoreSettings} containing information about the savepoint to restore from
+	 * @throws Exception if the {@link ExecutionGraph} could not be restored
+	 */
+	private void tryRestoreExecutionGraphFromSavepoint(ExecutionGraph executionGraphToRestore, SavepointRestoreSettings savepointRestoreSettings) throws Exception {
+		if (savepointRestoreSettings.restoreSavepoint()) {
+			final CheckpointCoordinator checkpointCoordinator = executionGraphToRestore.getCheckpointCoordinator();
+			if (checkpointCoordinator != null) {
+				checkpointCoordinator.restoreSavepoint(
+					savepointRestoreSettings.getRestorePath(),
+					savepointRestoreSettings.allowNonRestoredState(),
+					executionGraphToRestore.getAllVertices(),
+					userCodeLoader);
+			}
+		}
+	}
+
+	@Override
+	public void setMainThreadExecutor(final ComponentMainThreadExecutor mainThreadExecutor) {
+		this.mainThreadExecutor = checkNotNull(mainThreadExecutor);
+		executionGraph.start(mainThreadExecutor);
+	}
+
+	@Override
+	public void registerJobStatusListener(final JobStatusListener jobStatusListener) {
+		executionGraph.registerJobStatusListener(jobStatusListener);
+	}
+
+	@Override
+	public void startScheduling() {
+		mainThreadExecutor.assertRunningInMainThread();
+
+		try {
+			executionGraph.scheduleForExecution();
+		}
+		catch (Throwable t) {
+			executionGraph.failGlobal(t);
+		}
+	}
+
+	@Override
+	public void suspend(Throwable cause) {
+		mainThreadExecutor.assertRunningInMainThread();
+		executionGraph.suspend(cause);
+	}
+
+	@Override
+	public void cancel() {
+		mainThreadExecutor.assertRunningInMainThread();
+		executionGraph.cancel();
+	}
+
+	@Override
+	public CompletableFuture<Void> getTerminationFuture() {
+		return executionGraph.getTerminationFuture().thenApply(FunctionUtils.nullFn());
+	}
+
+	@Override
+	public boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
+		mainThreadExecutor.assertRunningInMainThread();
+		return executionGraph.updateState(taskExecutionState);
+	}
+
+	@Override
+	public SerializedInputSplit requestNextInputSplit(JobVertexID vertexID, ExecutionAttemptID executionAttempt) throws IOException {
+		mainThreadExecutor.assertRunningInMainThread();
+
+		final Execution execution = executionGraph.getRegisteredExecutions().get(executionAttempt);
+		if (execution == null) {
+			// can happen when JobManager had already unregistered this execution upon on task failure,
+			// but TaskManager get some delay to aware of that situation
+			if (log.isDebugEnabled()) {
+				log.debug("Can not find Execution for attempt {}.", executionAttempt);
+			}
+			// but we should TaskManager be aware of this
+			throw new IllegalArgumentException("Can not find Execution for attempt " + executionAttempt);
+		}
+
+		final ExecutionJobVertex vertex = executionGraph.getJobVertex(vertexID);
+		if (vertex == null) {
+			log.error("Cannot find execution vertex for vertex ID {}.", vertexID);
```

Comment:
I think we could remove the error logging statements and log them on the `JobMaster` level.

Reply:
After the `throw`, we potentially leave the _distributed component_, which means we should also catch `IllegalArgumentException` in `JobMaster#requestNextInputSplit()` to log the exception before returning an exceptional future. It would be consequent to catch all `RuntimeExceptions` on `JobMaster` side before leaving the _distributed component_, which is evidently not done for most of the methods. Why not log exceptional futures in the rpc framework?